File: wrappers/dotnet/src/Hyperledger.Ursa.BbsSignatures/Keys/BlindedBlsKeyPair.cs
using System.Text;

namespace BbsSignatures
{
    public class BlindedBlsKeyPair : BlsKeyPair
    {
        /// <summary>
        /// Default BLS 12-381 private key length
        /// </summary>
        public static int BlindingFactorSize => NativeMethods.blinding_factor_size();

        public BlindedBlsKeyPair(byte[] publicKey, byte[] secretKey, byte[] blindingFactor)
            : base(publicKey, secretKey)
        {
            BlindingFactor = blindingFactor;
        }

        public byte[] BlindingFactor { get; internal set; }

        public static new BlindedBlsKeyPair GenerateG1(string? seed = default)
        {
            using var context = new UnmanagedMemory();

            NativeMethods.bls_generate_blinded_g1_key(
                seed is null ? ByteBuffer.None : context.ToBuffer(Encoding.UTF8.GetBytes(seed)),
                out var publicKey, out var secretKey,
                out var blindingFactor,
                out var error);
            context.ThrowOnError(error);

            return new BlindedBlsKeyPair(
                publicKey: context.ToByteArray(publicKey),
                secretKey: context.ToByteArray(secretKey),
                blindingFactor: context.ToByteArray(blindingFactor));
        }

        public static new BlindedBlsKeyPair GenerateG2(string? seed = default)
        {
            using var context = new UnmanagedMemory();

            NativeMethods.bls_generate_blinded_g2_key(
                seed is null ? ByteBuffer.None : context.ToBuffer(Encoding.UTF8.GetBytes(seed)),
                out var publicKey,
                out var secretKey,
                out var blindingFactor,
                out var error);
            context.ThrowOnError(error);

            return new BlindedBlsKeyPair(
                publicKey: context.ToByteArray(publicKey),
                secretKey: context.ToByteArray(secretKey),
                blindingFactor: context.ToByteArray(blindingFactor));
        }
    }
}

File: wrappers/python/ursa_bbs_signatures/models/keys/BlindedBlsKeyPair.py
from typing import Optional, Union

from ..._ffi.bindings.bls import (
    bls_generate_blinded_g1_key,
    bls_generate_blinded_g2_key, blinding_factor_size,
)
from ..._ffi.ffi_util import encode_bytes
from .BlsKeyPair import BlsKeyPair


class BlindedBlsKeyPair(BlsKeyPair):
    def __init__(self, public_key: bytes, secret_key: bytes, blinding_factor: bytes):
        super().__init__(public_key, secret_key=secret_key)
        self.blinding_factor = blinding_factor

    @staticmethod
    def blinding_factor_size() -> int:
        return blinding_factor_size()

    @classmethod
    def generate_g1(
        cls, seed: Optional[Union[str, bytes]] = None
    ) -> "BlindedBlsKeyPair":
        res = bls_generate_blinded_g1_key(seed)
        return cls(res["public_key"], res["secret_key"], res["blinding_factor"])

    @classmethod
    def generate_g2(
        cls, seed: Optional[Union[str, bytes]] = None
    ) -> "BlindedBlsKeyPair":
        res = bls_generate_blinded_g2_key(seed)
        return cls(res["public_key"], res["secret_key"], res["blinding_factor"])
